### DISCUSSION TITLE
Minio with nginx and https not working

### DISCUSSION THREAD

**Discussion by github-primsec** [ente-io/ente#2546] [Q&A] [answered]
Hello!

I have the following problem:

I have created ente photos as a container and also created the webapp as a container. Basically, both ente and the web app work.
However, I always see empty pictures on the website after logging in (same when I share an album publicly). When troubleshooting via the website, I came across the error “Mixed content”, as the communication via the website and to Museum is secured with https and also works, but it then wants to access the Minio backend (port 3200) directly, which in turn is currently http.

Ich wollte entsprechend Minio hinter dem nginx packen und ebenfalls per https absichern, bekomme dann jedoch die folgende Meldung:

```
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
<Key>1580559962386438/ecedb96b-1733-44c2-9581-bd986def75a9</Key>
<BucketName>b2-eu-cen</BucketName>
<Resource>/b2-eu-cen/1580559962386438/ecedb96b-1733-44c2-9581-bd986def75a9</Resource>
<RequestId>17E56896622D1837</RequestId>
<HostId>3f2e9c5dc7d653a716cab27333b1b1f325cb9058c238c752500a3c61eea26b00</HostId>
</Error>
```

If I access the page directly (i.e. minio via http) I get the file delivered in binary form. What do I have to enter in nginx so that the forwarding to minio works correctly or does it not work at all?

I have tried it with:

```
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # WebSocket support for MinIO
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection “upgrade”;
```

but unfortunately unsuccessful.

**Comment by mnvr** [comment]
You have diagnosed the problem correctly. You need to configure nginx to terminate the HTTPS connection (or figure out some other way of directly running the minio over HTTPS). If it helps, here is a sample nginx conf that we use - https://github.com/ente-io/ente/blob/main/server/scripts/deploy/museum.nginx.conf#L16 - this is not for minio though, it's for museum, but it has HTTPS on so maybe it provides some hint.

About the "The request signature we calculated does not match the signature you provided. ", you also need to make sure _both_ museum and your client are connecting over HTTPS, and to the same public URL. If museum connects to minio over a different URL than what the client uses, then the signature of the presigned URL will not match.

**Comment by github-primsec** [comment]
Thanks for your support. I think my main issue was that I don´t setup museum to the same url like the web site. Museum was using the "old" path to ip+Port 3200.

But know I´m getting a cors error:

"...has been blocked by CORS policy: Request header field cache-control is not allowed by Access-Control-Allow-Headers in preflight response."

This is comming from the ente endpoint (museum) to the web app. I tested it with this in the location block of nginx:

```
        add_header 'Access-Control-Allow-Origin' "*" always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Headers' "Origin, X-Requested-With, Content-Type, Accept" always;
        add_header 'Access-Control-Allow-Methods' '*';
        add_header Cache-Control "public, no-transform";
        add_header 'Access-Control-Max-Age' 1728000;
        add_header 'Content-Type' 'text/plain; charset=utf-8';
        add_header 'Content-Length' 0;
        return 204;
        }
```
        
But it is still not working.

**Reply by mnvr** [reply]
You're adding the `Access-Control-Allow-Headers` in nginx. This is overriding the `Access-Control-Allow-Headers` returned by museum. The `Access-Control-Allow-Headers` returned by museum (https://github.com/ente-io/ente/blob/f0d5c7d51f47379e1d12a82ef44ef1cf92c6c7cb/server/cmd/museum/main.go#L930) include `Cache-Control`, but the override in nginx is removing it from the list of allowed headers, thus the request fails.

**Reply by github-primsec** [reply]
I removed all headers from nginx for museum and web but still the same issue:

`from origin 'https://*****' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

**Comment by github-primsec** [comment]
I also tried to add all the headers above in nginx location block:

```
...
                if ($request_method = OPTIONS) {
                add_header 'Access-Control-Allow-Origin' '*';
                add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, PUT, DELETE';
                add_header 'Access-Control-Allow-Headers' 'Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, X-Auth-Token, X-Auth-Access-Token, X-Cast-Access>
                add_header 'Access-Control-Max-Age' 86400;
                return 204;
                }
        add_header 'Access-Control-Allow-Origin' '*';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, PUT, DELETE';
        add_header 'Access-Control-Allow-Headers' 'Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, X-Auth-Token, X-Auth-Access-Token, X-Cast-Access-Token, >

```
        }

but still the same issue

**Comment by github-primsec** [comment, marked as answer]
Its now working...

I had two issues... first (because of a "tab" in the config musem was crashing... so no backend was up :/)
second:

I setup the following in nginx (location path)

                if ($request_method = OPTIONS) {
                add_header 'Access-Control-Allow-Origin' '**https://****.****';
                add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS, PUT, DELETE';
                add_header 'Access-Control-Allow-Headers' 'Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, X-Auth-Token, X-Auth-Access-Token, X-Cast-Access>
                add_header 'Access-Control-Max-Age' 86400;
                return 204;
                }

I saw the error that he don´t allow these Headers two times and also not with '*'. I need adress the domain directly
